File: app/src/main/java/eu/kanade/tachiyomi/ui/base/presenter/NucleusConductorDelegate.java
package eu.kanade.tachiyomi.ui.base.presenter;

import android.os.Bundle;
import android.support.annotation.Nullable;

import nucleus.factory.PresenterFactory;
import nucleus.presenter.Presenter;

public class NucleusConductorDelegate<P extends Presenter> {

    @Nullable private P presenter;
    @Nullable private Bundle bundle;

    private PresenterFactory<P> factory;

    public NucleusConductorDelegate(PresenterFactory<P> creator) {
        this.factory = creator;
    }

    public P getPresenter() {
        if (presenter == null) {
            presenter = factory.createPresenter();
            presenter.create(bundle);
            bundle = null;
        }
        return presenter;
    }

    Bundle onSaveInstanceState() {
        Bundle bundle = new Bundle();
//        getPresenter(); // Workaround a crash related to saving instance state with child routers
        if (presenter != null) {
            presenter.save(bundle);
        }
        return bundle;
    }

    void onRestoreInstanceState(Bundle presenterState) {
        bundle = presenterState;
    }

    void onTakeView(Object view) {
        getPresenter();
        if (presenter != null) {
            //noinspection unchecked
            presenter.takeView(view);
        }
    }

    void onDropView() {
        if (presenter != null) {
            presenter.dropView();
        }
    }

    void onDestroy() {
        if (presenter != null) {
            presenter.destroy();
        }
    }
}

File: app/src/main/java/eu/kanade/tachiyomi/ui/base/presenter/NucleusConductorDelegate.kt
package eu.kanade.tachiyomi.ui.base.presenter

import android.os.Bundle
import nucleus.factory.PresenterFactory
import nucleus.presenter.Presenter

class NucleusConductorDelegate<P : Presenter<*>>(private val factory: PresenterFactory<P>) {

    var presenter: P? = null
        get() {
            if (field == null) {
                field = factory.createPresenter()
                field!!.create(bundle)
                bundle = null
            }
            return field
        }

    private var bundle: Bundle? = null

    fun onSaveInstanceState(): Bundle {
        val bundle = Bundle()
        //        getPresenter(); // Workaround a crash related to saving instance state with child routers
        presenter?.save(bundle)
        return bundle
    }

    fun onRestoreInstanceState(presenterState: Bundle?) {
        bundle = presenterState
    }

    @Suppress("UNCHECKED_CAST")
    private fun <View> Presenter<View>.takeView(view: Any) = takeView(view as View)

    fun onTakeView(view: Any) {
        presenter?.takeView(view)
    }

    fun onDropView() {
        presenter?.dropView()
    }

    fun onDestroy() {
        presenter?.destroy()
    }
}
